### PR TITLE
add label to identify individual services

### PR DESCRIPTION
**Pull request by zjj2wry**
### What problem does this PR solve? 
Fixes https://github.com/pingcap/tidb-operator/issues/2622

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
```bash
➜  tidb-operator git:(add-labels-for-service) ✗ kubectl get svc -n pingcap-dev
NAME                     TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)              AGE
basic-discovery          ClusterIP   10.107.212.188   <none>        10261/TCP            2m3s
basic-grafana            ClusterIP   10.111.254.210   <none>        3000/TCP             2m2s
basic-monitor-reloader   ClusterIP   10.97.149.92     <none>        9089/TCP             2m2s
basic-pd                 ClusterIP   10.110.120.114   <none>        2379/TCP             2m2s
basic-pd-peer            ClusterIP   None             <none>        2380/TCP             2m2s
basic-prometheus         ClusterIP   10.103.60.68     <none>        9090/TCP             2m3s
basic-tidb               ClusterIP   10.110.92.244    <none>        4000/TCP,10080/TCP   12s
basic-tidb-peer          ClusterIP   None             <none>        10080/TCP            12s
basic-tikv-peer          ClusterIP   None             <none>        20160/TCP            60s
➜  tidb-operator git:(add-labels-for-service) ✗ kubectl get svc -n pingcap-dev -l app.kubernetes.io/used-by=end-user
NAME         TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)              AGE
basic-pd     ClusterIP   10.110.120.114   <none>        2379/TCP             2m15s
basic-tidb   ClusterIP   10.110.92.244    <none>        4000/TCP,10080/TCP   25s
➜  tidb-operator git:(add-labels-for-service) ✗ kubectl get svc -n pingcap-dev -l app.kubernetes.io/used-by=internal
NAME              TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)     AGE
basic-pd-peer     ClusterIP   None         <none>        2380/TCP    2m21s
basic-tidb-peer   ClusterIP   None         <none>        10080/TCP   31s
basic-tikv-peer   ClusterIP   None         <none>        20160/TCP   79s
➜  tidb-operator git:(add-labels-for-service) ✗ kubectl get svc -n pingcap-dev -l app.kubernetes.io/used-by=grafana
NAME            TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)    AGE
basic-grafana   ClusterIP   10.111.254.210   <none>        3000/TCP   2m28s
➜  tidb-operator git:(add-labels-for-service) ✗ kubectl get svc -n pingcap-dev -l app.kubernetes.io/used-by=prometheus
NAME               TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)    AGE
basic-prometheus   ClusterIP   10.103.60.68   <none>        9090/TCP   2m35s
```

Code changes

 - Has Go code change

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

**Comment by zjj2wry** [comment]
@cofyc @Yisaer this pr ready for review

File: pkg/label/label.go [hunk 1]
```diff
@@ -24,6 +24,11 @@ import (
 const (
 	// The following labels are recommended by kubernetes https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/
 
+	// ResourceNameLabelKey is the unique identifier of the resource, its value is the same as metadata.name
+	ResourceNameLabelKey string = "kubernetes.io/name"
+	// UsedByLabelKey indicate where it is used. for example, tidb has two services,
+	// one for internal component access and the other for end-user
+	UsedByLabelKey string = "app.kubernetes.io/used-by"
```
Comment (cofyc) [inline pkg/label/label.go:31]: move our labels after `NamespaceLabelKey`

ManagedByLabelKey, ComponentLabelKey, etc. are well-known labels recommended [here](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels) (see above comments)

File: pkg/monitor/monitor/util_test.go [hunk 2]
```diff
@@ -51,6 +51,8 @@ func TestGetMonitorService(t *testing.T) {
 							"app.kubernetes.io/managed-by": "tidb-operator",
 							"app.kubernetes.io/instance":   "foo",
 							"app.kubernetes.io/component":  "monitor",
+							"app.kubernetes.io/used-by":    "prometheus",
+							"kubernetes.io/name":           "foo-prometheus",
```
Comment (Yisaer) [inline pkg/monitor/monitor/util_test.go:55]: I find this test haven't covered the grafana. Could you extend this unit-test to cover grafana?

**Comment by cofyc** [comment]
@kolbe you requested this feature, can you help review this PR?

File: pkg/label/label.go [hunk 3]
```diff
@@ -24,6 +24,11 @@ import (
 const (
 	// The following labels are recommended by kubernetes https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/
 
+	// ResourceNameLabelKey is the unique identifier of the resource, its value is the same as metadata.name
+	ResourceNameLabelKey string = "kubernetes.io/name"
```
Comment (cofyc) [inline pkg/label/label.go:28]: if we have a label to identify resources by purposes, is this label necessary? users can always get resources by name.

Comment (zjj2wry) [reply pkg/label/label.go]: will only used by user want use label identify a service, but this doesn’t necessary. deleted

File: pkg/monitor/monitor/util_test.go (hunk 2 above)
Comment (zjj2wry) [reply pkg/monitor/monitor/util_test.go]: of course, done

File: pkg/manager/member/tikv_member_manager_test.go [hunk 4]
```diff
@@ -1503,6 +1503,8 @@ func TestGetNewTiFlashServiceForTidbCluster(t *testing.T) {
 						"app.kubernetes.io/managed-by": "tidb-operator",
 						"app.kubernetes.io/instance":   "foo",
 						"app.kubernetes.io/component":  "tikv",
+						"app.kubernetes.io/used-by":    "internal",
```
[the remark this answers is not in the document]
Comment (zjj2wry) [reply pkg/manager/member/tikv_member_manager_test.go]: make sense, done

File: pkg/label/label.go (hunk 1 above)
Comment (zjj2wry) [reply pkg/label/label.go]: done

**Comment by tirsen** [comment]
LGTM

File: pkg/manager/member/tiflash_member_manager_test.go [hunk 5]
```diff
@@ -1186,6 +1186,7 @@ func TestGetNewServiceForTidbCluster(t *testing.T) {
 						"app.kubernetes.io/managed-by": "tidb-operator",
 						"app.kubernetes.io/instance":   "foo",
 						"app.kubernetes.io/component":  "tiflash",
+						"app.kubernetes.io/used-by":    "peer",
```
Comment (DanielZhangQD) [inline pkg/manager/member/tiflash_member_manager_test.go:1189]: Please also help update the labels for Pump, TiFlash and TiCDC.

**Comment by tirsen** [comment]
`app.kubernetes.io/used-by` doesn't look like it's a standard label? Are there any docs for this label?

**Comment by zjj2wry** [comment]
> `app.kubernetes.io/used-by` doesn't look like it's a standard label? Are there any docs for this label?

yes, it is a custom define label. kubernetes recommended label will used for all resources, this will only used for service. thanks

File: pkg/label/label.go [hunk 6]
```diff
@@ -223,6 +225,24 @@ func (l Label) Instance(name string) Label {
...
 }
 
+// UserBy adds use-by kv pair to label
+func (l Label) UsedBy(name string) Label {
+	l[UsedByLabelKey] = name
+	return l
+}
+
+// UsedByInternal adds used-by=internal label
+func (l Label) UsedByInternal() Label {
```
[the remark this answers is not in the document]
Comment (zjj2wry) [reply pkg/label/label.go]: done

File: pkg/manager/member/tiflash_member_manager_test.go (hunk 5 above)
Comment (zjj2wry) [reply pkg/manager/member/tiflash_member_manager_test.go:1189]: Pump, TiFlash and TiCDC also have multiply service, this label will only by when a component have mulitply service

Comment (DanielZhangQD) [reply pkg/manager/member/tiflash_member_manager_test.go:1189]: ok, got it.

**Pull request merged**